File: introductionToAlgorithmCLRS/chp6/Exercises/Ex_6.2/q2MinHeapify.java
public class q2MinHeapify {
    public static int Parent(int i) {
        return ((i-1)/2);
    }

    public static int Left(int i) {
        return ((2*i)+1);
    }

    public static int Right(int i) {
        return ((2*i)+2);
    }

    public static void Min(int array[], int i, int size) {

        int l = Left(i);
        int r = Right(i);
        int largest;

        if (l<size && array[l]<array[i]) {
            largest = l;
        }
        else {
            largest = i;
        }
        
        if (largest != i) {
            int temp = array[i];
            array[i] = array[largest];
            array[largest] = temp;
            Min(array, largest, size);
        }
        if (r < size && array[r]<array[i]) {
            largest = r;
        }
        else {
            largest = i;
        }
        

        if (largest != i) {
            int temp = array[i];
            array[i] = array[largest];
            array[largest] = temp;
            Min(array, largest, size);
        }
    }
}

File: introductionToAlgorithmCLRS/chp6/Exercises/Ex_6.2/q2MinHeapify.c
int Parent(int i) {
    return (i/2);
}

int Left(int i) {
    return (2*i)+1;
}

int Right(int i) {
    return (2*i)+2;
}

void Max (int array[], int i, int heapSize) {

    int l = Left(i);
    int r = Right(i);
    int smallest;

    if (l<=heapSize && array[l] < array[i]) {
        smallest = l;
    }
    else {
        smallest = i;
    }

    if (r <= heapSize && array[r] < array[smallest]) {
        smallest = r;
    }

    if (smallest != i) {
        int temp = array[i];
        array[i] = array[smallest];
        array[smallest] = temp;
        Max(array, smallest, heapSize);
    }
}

